### PR TITLE
New Fluent expander style

### LINKED ISSUE #5358: Port new Expander Fluent style from WinUI
See:
https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Expander/Expander_themeresources_v2.5.xaml
https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Expander/Expander.xaml

image

(on this screenshot header font size is not part of default fluent template)

### PR DESCRIPTION
**Pull request by Splitwirez**
## What does the pull request do?
Updates the appearance of the `Expander` control to match that of WinUI 3 when using the Avalonia Fluent theme.
Avalonia-NewFluentExpanderStyle

## What is the current behavior?
`Expander` looks nothing like its WinUI 3 counterpart, and looks completely out-of-place next to other built-in controls when using the Avalonia Fluent theme.

## What is the updated/expected behavior with this PR?
The appearance of the `Expander` control will now match that of its WinUI 3 counterpart when using the Avalonia Fluent theme.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
I guess this could mess with any apps which tweak or heavily depend on the existing `Expander` style (though I'd hardly be surprised if some of such apps do so for the same reason for which this PR exists

## Fixed issues
Fixes #5358

### CODE REVIEW

**Comment by maxkatz6** [comment]
Thanks! Looks good to me, but I noticed one small problem.

In Left/Right mode header height is increasing when control expands with simple one-line text content in both header and content. Because of that layout visually jumps for a little bit, even on your gif with "Left" expander.
It's probably because of margins/paddings inconsistency in WinUI control, while it's not noticeable there without left/right modes.
I don't think if it's really that important, but if you have some idea, it would be nice.

**Comment by Splitwirez** [comment]
@maxkatz6 Done and done :+1:

**Review by maxkatz6 — DISMISSED** [review]
As for me, everything looks good.
Not sure about "breaking changes" part.
It definitely will break some styles at least in one my app, but I don't think it's critical.

There are some another nits about naming, but it's fine.

File: src/Avalonia.Themes.Fluent/Controls/Expander.xaml
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/Expander.xaml:46]: It's preferred to use "PART_" naming in default styles. Here and above.

Comment (Splitwirez) [reply src/Avalonia.Themes.Fluent/Controls/Expander.xaml:46]: I thought the "PART_" prefix was only for elements which are accessed from the control's code, such as in the `OnApplyTemplate` function...have I been getting that wrong this whole time?

**Comment by robloo** [comment]
Just a few very minor observations:

 1. The WinUI expander does not have rounded corners between the header and content when in the expanded state. The header and content should appear as a single rectangular region with 4 rounded corners on the outside edges only.
 2. I'm not sure the header/content background matches the WinUI style.

image

File: src/Avalonia.Themes.Fluent/Controls/Expander.xaml
Comment (robloo) [reply src/Avalonia.Themes.Fluent/Controls/Expander.xaml:46]: For UWP/WPF: "PART_ExpanderContent" should only be used if the control is referenced in code-behind. Standard naming convention ("ExpanderContent") should be used for controls used by name in the template/style only. Not sure if Avalonia convention is different for their styling system.

Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/Expander.xaml:46]: Well, as for me there are no a lot of difference between controls used in OnApplyTemplate and controls that can be potentially used there in some day or just additional components of layout. Especially when they all can be used in the styles without OnApplyTemplate. And personally I used "PART_" everywhere for consistency,

On other hand, as I mentioned above, it's not a blocker for me.

**Comment by Splitwirez** [comment]
> Just a few very minor observations:
> 
>     1. The WinUI expander does not have rounded corners between the header and content when in the expanded state. The header and content should appear as a single rectangular region with 4 rounded corners on the outside edges only.
> 
>     2. I'm not sure the header/content background matches the WinUI style.
> 
> 
> image

Funny...reading the XAML, I thought that must be the case, but the only images I could find indicated otherwise. Guess they must've been outdated...I'll have to fix that.

**Comment by maxkatz6** [comment]
@Splitwirez Background and Border changes come from new WinUI 3 styles, which is not yet released. That's why you have seen old styles. 

About corner radius, IIRC they used converter with same ControlCornerRadius resource, where that converter is zeroing unwanted corners. In Controls project we already have similar converter, but only for margins/paddings.

**Comment by robloo** [comment]
@Splitwirez Hm, well the image above is from the latest released preview: 2.6.0-prerelease.210217002 (released February 17, 2021). It is NOT from the WinUI 3 version of this control.

They just recently did another round of visual updates here https://github.com/microsoft/microsoft-ui-xaml/pull/4159 Visual styles seem to match what I see in my apps which do not yet use WinUI 3.0. Perhaps you just didn't have the latest code? I haven't compiled from source recently to check this myself.

File: src/Avalonia.Themes.Fluent/Controls/Expander.xaml
Comment (robloo) [reply src/Avalonia.Themes.Fluent/Controls/Expander.xaml:46]: Well, the "PART_" naming simply means it is a part of the template referenced or required by code. A style/template may have to name any number of controls for styling reasons but most don't need to be referenced outside the style itself. The "PART_" is to indicate a link from code -> style NOT from style -> style (which is arbitrary and the code doesn't need to know).

Comment (Splitwirez) [reply src/Avalonia.Themes.Fluent/Controls/Expander.xaml:46]: If it's not a blocker, then I think I'll continue to reserve "PART_" for the bits used in code, as I previously have (and as robloo noted is standard practice in WPF and UWP). Working on the corners as we speak.

**Comment by Splitwirez** [comment]
Avalonia-NewFluentExpanderStyle-v2
How's this?

File: src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: I believe it will be better to port the same converter as WinUI has, because it's not wired to the specific control and can be reused in the Avalonia and be the users:

https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Common/CornerRadiusFilterConverters.idl
https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Common/CornerRadiusFilterConverter.h
https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Common/CornerRadiusFilterConverter.cpp

**Review by maxkatz6 — CHANGES_REQUESTED** [review]

**Comment by robloo** [comment]
FYI, here is the current control from the WinUI 2.x library (built just now as I look to fix some upstream issues). There is currently only one style for this control regardless of WinUI 2/3.

image

File: src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: Oh, so *that's* where those converters live in MS's repo...I'll get on that ASAP.


Honestly, this isn't at all how I would address this round corner problem, but...that's another topic, for another Issue/PR/etc.

Comment (Splitwirez) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: Update: Halfway through porting the WinUI converters, I realized something: These converters operate under the assumption that most (if not all) controls have a `CornerRadius` property.

This assumption seems to be true for WinUI, but is not so for Avalonia.


However, I'm not convinced that following WinUI's example (e.g. adding a `CornerRadius` to every control) is the way to go for Avalonia.

I'll elaborate further in an Issue post shortly, since it's a separate topic, one which may have ramifications well beyond the `Expander` control.

Comment (Splitwirez) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: https://github.com/AvaloniaUI/Avalonia/discussions/5594 thoughts?

Comment (maxkatz6) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: @Splitwirez
> These converters operate under the assumption that most (if not all) controls have a CornerRadius property.

Why though? It's simple IValueConverter that has CornerRadius as an input and has some filter/scale properties with returning processed value.

Comment (maxkatz6) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: Discussion about adding a CornerRadius to Avalonia controls is a good idea.

Comment (robloo) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: TopLeftValue and BottomRightValue filter kind are strange in WinUI. It doesn't make sense to me right away what these do (ill admit I didnt check the code eithet). If might be best to remove these for Avalonia though.

Comment (Splitwirez) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: > Why though? It's simple IValueConverter that has CornerRadius as an input and has some filter/scale properties with returning processed value. It doesn't need to operate with these controls directly.

@maxkatz6 Well yes...but that CornerRadius input has to come from somewhere. If not the templated parent, then...where?

Comment (Splitwirez) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: > Discussion about adding a CornerRadius to Avalonia controls is a good idea.

As explained in my discussion post...while I don't like that idea personally, I think it will probably meet the needs of most developers better than my proposed alternative. Once a final decision is reached, I'd be more than happy to help implement it...even if that means just a `CornerRadius` property and corresponding style/template changes. I guess that should be a separate PR from this one though, correct?

**Comment by Splitwirez** [comment]
@robloo I forgot to say this the other day, but...thank you for providing this known good reference image.

File: src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs
Comment (maxkatz6) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: > If not the templated parent, then...where?

In WinUI it was done with ThemeResource on older platforms. 
```<Border CornerRadius={Binding Source={StaticResource ControlCornerRadius}, Converter={StaticResource Converter}} />```
I am not sure if DynamicResource will work there though, as it would be preferable.

Comment (maxkatz6) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: >  I guess that should be a separate PR from this one though, correct?

Yes, "CornerRadius for all controls or something" related questions should be outside of this PR.

Comment (Splitwirez) [reply src/Avalonia.Controls/Converters/BottomCornerRadiusFilterConverter.cs:7]: @maxkatz6 Agreed. I've reached out to you on Discord to discuss how to proceed.

**Comment by robloo** [comment]
FYI, I made several upstream changes to the default style to fix and improve things. I'm not sure when that will get merged in WinUI but ideally those changes will follow through here.

**Comment by Splitwirez** [comment]
> FYI, I made several upstream changes to the default style to fix and improve things. I'm not sure when that will get merged in WinUI but ideally those changes will follow through here.

tbh I may just end up scrapping this PR, helping do whatever is needed for the `CornerRadius` stuff, and then redo this style port once that's merged into Avalonia and the WinUI Expander style changes are also merged into...well, WinUI

**Comment by robloo** [comment]
@Splitwirez 

I didn't mean to take the wind out of your sails for this PR. I think it would be great to merge this PR first as-is (only waiting on the CornerRadius converter updates I think). A later PR by myself or someone else could then update to the latest WinUI style after it's merged upstream.

**Comment by Splitwirez** [comment]
> I didn't mean to take the wind out of your sails for this PR. 

@robloo No no, it wasn't your fault! I had already figured I would go the route I described before your comment about your PR to WinUI.

> I think it would be great to merge this PR first as-is (only waiting on the CornerRadius converter updates I think). A later PR by myself or someone else could then update to the latest WinUI style after it's merged upstream.

Hmmm...maybe you're right...yeah, I guess I may as well. For all I know, I might end up porting your changes myself if they get merged into WinUI. We'll see, lol

**Comment by Splitwirez** [comment]
@maxkatz6 I understand you wanted me to implement the `BottomCornerRadiusFilterConverter` and `TopCornerRadiusFilterConverter` as they are in WinUI, but...the links you provided only point to some other `CornerRadiusFilterConverter` class (which doesn't even implement `IValueConverter`)...in fact, I've searched the entire repo and found no mention of `BottomCornerRadiusFilterConverter` or `TopCornerRadiusFilterConverter` outside of XAML...any idea what's up with that, or what I ought to do in response?

**Comment by robloo** [comment]
> @maxkatz6 I understand you wanted me to implement the `BottomCornerRadiusFilterConverter` and `TopCornerRadiusFilterConverter` as they are in WinUI, but...the links you provided only point to some other `CornerRadiusFilterConverter` class (which doesn't even implement `IValueConverter`)...in fact, I've searched the entire repo and found no mention of `BottomCornerRadiusFilterConverter` or `TopCornerRadiusFilterConverter` outside of XAML...any idea what's up with that, or what I ought to do in response?

I can answer that. There is only one 'converter' here and it's called CornerRadiusFilterConverter. However, it is a bit specialized in a way you might not expect. It has an additional `Filter` and `Scale` property and also an additional `Convert` method (along with the standard Convert/ConvertBack).

In WinUI you can never use a converter directly and have to first declare it as a resource in XAML with a key. Microsoft did that and created the `BottomCornerRadiusFilterConverter` and the `TopCornerRadiusFilterConverter`. But these converters also specify the Filter kind so they can be used with bindings directly (with no parameters as might be more common).

I'm just started to dip my toes into Avalonia so don't know if converters need to be declared in a different way (I suspect you might be able to use them directly without a key). @maxkatz6 Would need to help there.

I also wouldn't recommend including the CornerRadiusFilterKind.TopLeftValue or CornerRadiusFilterKind.BottomRightValue. This just muddies things up a bit and their usage is not clear.

Finally, as an FYI, my Expander changes have been merged upstream.

**Comment by maxkatz6** [comment]
@Splitwirez it's definitely harder to see familiar converter structure from C++/WinRT code, but as @robloo said, that converter in WinUI repo has all the methods required to implement IValueConverter, and also has properties that allows to configure converter instances to work with specific corners. Lately these instances are stored in the WinUI resources (named as "BottomCornerRadiusFilterConverter" as other).

> I'm just started to dip my toes into Avalonia so don't know if converters need to be declared in a different way (I suspect you might be able to use them directly without a key). @maxkatz6 Would need to help there.

It's possible to declare static instances of these converters and use them with {x:Static} markup extension. 
In Avalonia we already have some default converters defined like that: https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Base/Data/Converters

But it doesn't allow to specific typed parameters with this markup extension, so these converters should be prepared for specific cases as well, as in UWP/WPF. Although there is ConverterParameter... 

I don't think that we need 100% matching API for this converter comparing with WinUI, there is not a lot of advantages for doing it. But since this converter will be part of Avalonia API after it's merged, it should be useful for other controls and developers as well.

**Comment by robloo** [comment]
> But it doesn't allow to specific typed parameters with this markup extension, so these converters should be prepared for specific cases as well, as in UWP/WPF. Although there is ConverterParameter...

If it was me, I would use the converter parameter to pass the filter kind. This wouldn't be quite as performant as you would then be parsing in the converter... although you could just use the string directly.

**Review by maxkatz6 — APPROVED** [review]

**Comment by maxkatz6** [comment]
Technically, it's expander style breaking change, if anybody has custom one.
If user wants to have this style right now, he can just copy paste it.

**Pull request merged**
